packages/babel-parser/src/parser/util.js

```diff
@@ -273,24 +283,26 @@ export default class UtilParser extends Tokenizer {
   checkExpressionErrors(
     refExpressionErrors: ?ExpressionErrors,
     andThrow: boolean,
-  ) {
-    if (!refExpressionErrors) return false;
-    const { shorthandAssign, doubleProto, optionalParameters } =
+  ): void {
+    if (!andThrow || !refExpressionErrors) {
+      return;
```

Comment:
So, this was difficult to understand, but *I think* this is what's going on:

1. The intent, as I understand it, of `refExpressionErrors` is to "hold onto" things that would be errors in expressions, but *not* in patterns.
2. The only places patterns can appear is in variable declarations, function parameters, and assignment expressions (which are ambiguous until we reach an `=` sign).
3. In the first two, `refExpressionErrors` is just passed down as `null`, so it is irrelevant for our discussion here.
4. However, if it the current expression was determined to be an assignment expression in `parseMaybeAssign`, then `refExpressionErrors` will have already been cleared since we know this *was* a pattern ( https://github.com/babel/babel/blob/5b4d55d6e532e3c2cd7284fe951fc792108a62af/packages/babel-parser/src/parser/expression.js#L323 )
5. As such, I *think* the check in `parseUpdate` never passes, as by this point `refExpressionErrors` has been cleared. `parseUpdate` is the only place that passes `false` for `andThrow` (and also the only place that bothers checking the result).

This is why no tests fail. I've done my best to construct expressions that find some other path where this line would have `refExpressionErrors` and rely on andThrow being false to avoid throwing them, but I haven't been able to. If you can confirm the above, I can try to modify the code further to reflect how it's being used.

Reply:
Just want to add that I think it's totally possible that there is a relevant case that's handled by this and simply not tested, and I'm happy to change this code back to that, and also add said test if we can find it.

Reply:
OK I went ahead and restored the old behavior I think, although I still can't seem to figure out how it behaves differently, but figured it wasn't worth blocking this PR on figuring that out. I can file a bug on exploring this further. We should try to get a test in there or figure out why it's never entered.